### ISSUE TITLE
Code cells missing "outputs" and "execution_count" are not understood and deleted

### ISSUE THREAD

**Issue by prinsherbert** [jupyterlab/jupyterlab#9933]
### Problem

* I'm always frustrated when opening notebooks where code cells miss the `execution_count` or `outputs` fields because these and next cells are removed.
* I would like it if one of these fields is missing jupyter(hub) assumes `0` and `[]` respectively, and continues to load.

### Proposed Solution

Assume 0 and `[]` for malformed notebooks

### Additional context

I had a script that would process notebooks to teacher and student versions, and I cleared the outputs and execution_count for the students' versions. This is fine for Google Colab, but students failed to load them in jupyter.

I've solved this by clearing them to `0` and `[]`, rather than dropping the fields, but it would be nice for others these malformed notebooks are still read in a sensible manner, with a warning rather than an error. 

Or at least, do not modify the file if you don't understand it. Now the cells are also deleted from the source file, which is bad.

![notebooks](https://user-images.githubusercontent.com/70648554/110804326-5d705600-8280-11eb-83f3-cb9988448363.png)

**Comment by jasongrout** [comment]
Can you attach an example notebook that has a problem?

Likely this issue should be filed with the [`nbformat`](https://github.com/jupyter/nbformat/) repo, which implements the validation, and not JupyterLab (which is just reporting errors from nbformat).

**Comment by prinsherbert** [comment]
Example notebook:

    {
        "nbformat": 4,
        "nbformat_minor": 0, 
        "metadata": {"id": "1"}, 
        "cells":  [
            {"cell_type": "code", "metadata": {"id": "A"}, "source": ["# First cell, won't be deleted\n"], "outputs": [], "execution_count": 0},
            {"cell_type": "code", "metadata": {"id": "B"}, "source": ["# Second cell, misses outputs field but won't be deleted\n"]},
            {"cell_type": "code", "metadata": {"id": "C"}, "source": ["# Third cell, will be deleted\n"]}
        ]
    }

**Comment by prinsherbert** [comment]
It's not just reporting errors though, the cells are also deleted from the file. I think it's poor behavior that a program changes a file, particularly deletes potentially relevant things, just because it doesn't understand it.

This is the file afterwards, note that the third cell (and any cell after in general) is deleted:

        {
        "cells": [
        {
        "cell_type": "code",
        "execution_count": 0,
        "metadata": {
            "id": "A"
        },
        "outputs": [],
        "source": [
            "# First cell, won't be deleted\n"
        ]
        },
        {
        "cell_type": "code",
        "execution_count": null,
        "metadata": {
            "id": "B"
        },
        "outputs": [],
        "source": [
            "# Second cell, misses outputs field but won't be deleted\n"
        ]
        }
        ],
        "metadata": {
        "id": "1",
        "kernelspec": {
        "display_name": "Python 3",
        "language": "python",
        "name": "python3"
        },
        "language_info": {
        "codemirror_mode": {
            "name": "ipython",
            "version": 3
        },
        "file_extension": ".py",
        "mimetype": "text/x-python",
        "name": "python",
        "nbconvert_exporter": "python",
        "pygments_lexer": "ipython3",
        "version": "3.8.5"
        }
        },
        "nbformat": 4,
        "nbformat_minor": 1
        }

**Comment by bollwyvl** [comment]
Yep, the [v4 schema](https://github.com/jupyter/nbformat/blob/master/nbformat/v4/nbformat.v4.schema.json#L194) is very clear:
```yaml
"required": ["id", "cell_type", "metadata", "source", "outputs", "execution_count"],
```

`id` is new, and not adding a lot of value yet, but everything else has been present for [7 years](https://github.com/jupyter/nbformat/blame/c60cd150e702bf64c4dacc156a94f50d13474b9a/IPython/nbformat/v4/nbformat.v4.schema.json#L153).

If something doesn't pass validation, it isn't A Notebook. As a community, we have few tools that help us keep many clients, servers and tools working together, so i'd wager this would be `wontfix`ed if raise to nbformat...

**Comment by blois** [comment]
Colab is a bit more accepting of these missing fields but AFAIK will re-serialize these files as valid notebooks.

`id` field is now supported as well if you load a notebook with nbformat 4.5, though by default Colab is not yet generating that.

**Comment by jasongrout** [comment]
> I think it's poor behavior that a program changes a file, particularly deletes potentially relevant things, just because it doesn't understand it.

I think this is a valid point, though. Not sure if JLab or nbformat is changing the file, but it certainly shouldn't be saving a new value to the file by default

**Comment by prinsherbert** [comment]
It's still trying to parse the file, even includes the violating cell, but then ignores and discards all the others. It makes sense to have something like this to resolve broken notebooks:

    for cell in notebook["cells"]:
      cell['outputs'] = cell.get('outputs', [])
      cell['execution_count'] = cell.get('output', 0)

Optionally, a warning could be shown as well, above each cell.

**Comment by agoose77** [comment]
On the subject of how to do this in accordance with the notebook spec, if you set the execution count to `null`, and outputs to `[]`, it mirrors the functionality in Jupyter Notebook / Lab.

**Comment by prinsherbert** [comment]
> On the subject of how to do this in accordance with the notebook spec, if you set the execution count to `null`, and outputs to `[]`, it mirrors the functionality in Jupyter Notebook / Lab.

Absolutely agree, and this issue has been solved for me by doing so. I could only imagine that something more sensible can be done when they are missing.
